### DISCUSSION TITLE
tutorials/scripting/groups

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#85] [User-contributed notes]
# tutorials/scripting/groups

Groups in Godot work like tags in other software. You can add a node to as many groups as you want. Then, in code, you can use the SceneTree to: Get a list of nodes in a group., Call a method on al...

https://docs.godotengine.org/en/stable/tutorials/scripting/groups.html

**Comment by Crystalwarrior** [comment]
There is no way to obtain nodes only in your node's scene group yet.
A workaround for this is to iterate through your node's children first, and then use `is_in_group(StringName)` method, for example:
```py
	for child in self.find_children("*"):
		if child.is_in_group("enemy"):
			child.set_visible(false)
```
(You need to use `find_children("*")` to recursively iterate through its children, `get_children()` won't be enough as it will only iterate through THIS node's children and not any further children of the child nodes)

**Reply by GeminiSquishGames** [reply]
I'm interested to know the use case, I think I'm missing something. Currently we cannot have duplicate group StringNames, regardless if it's global or scene, so "enemy" can only be a global group or a scene group.

**Reply by Ieida** [reply]
> I'm interested to know the use case, I think I'm missing something. Currently we cannot have duplicate group StringNames, regardless if it's global or scene, so "enemy" can only be a global group or a scene group.

I'm creating a multiplayer game currently and I need to set the visibility of spawners and synchronizers to avoid players that are not in the same lobby. A great solution would be to get notified of a signal whenever a spawner is added to a group, which would need to be lobby specific and there isn't even a signal for it for some reason, or the other option would be to request spawners and synchronizers from the lobby which is just goofy and it means every node would need to know what lobby it's a part of.

**Reply by CreepyGnome** [reply]
>  to get notified of a signal whenever a spawner is added to a group, 

I assume you have code that is adding spawners to the group, so that seems like the place that could emit a signal.  You can put this signal in an Autoload for anything to connect to that needs it. Then when added the code doing the adding can emit the signal.  

> would need to be lobby specific

> means every node would need to know what lobby it's a part of

Does not seem like it needs to be lobby specific, or that every node needs to know what lobby it is part of. 

That is one of the nice things about the observer pattern and signals, its ability to decouple your code. Godot allows you to can make your own signals and emit them that way you do not have to be dependent on the core classes do not support your needs as every game is different in their needs in many ways, and it would be difficult for Godot to cater to every possible use case as that would bloat the engine.

**Reply by im-deepfriedwater** [reply]
I actually needed to fetch groups only within a node's children, appreciate the code snippet!

**Comment by WldGeese** [comment]
When creating a group I do not see the option to make it global, nor do I see anything about global groups in Project Settings. Is anybody able to help direct me?

**Reply by Piralein** [reply]
Which version of godot do you use?

The global group UI is only available in godot 4.3 and onwards.
The global groups tab in `Project Settings` can be found inside the `Globals` -> `Groups` tab.

**Reply by WldGeese** [reply]
Thank you so much - I was using 4.2 thinking that was the latest version since I very recently downloaded it. This will teach me to always actually check going forward. I downloaded 4.3 and see the global groups tab now.

**Comment by heroofhyla** [comment]
This new interface seems a lot better for adding one node to multiple groups, but more cumbersome for adding multiple nodes to one group. Back in 4.2 you could use the [manage groups](https://i.imgur.com/5Tr9gcE.png) dialog to quickly add multiple nodes to the same group all at once, but I can't find it anymore. Is this no longer possible with the new interface?

**Comment by yongan-game** [comment]
"Select one or more nodes in the Scene dock then click the add button with the + symbol" has been in the documentation since 4.0, yet you still can't actually select multiple nodes; the node dock tells you to "select a single node to edit its signals and groups" whenever you select more than one. Given that the Manage Groups option appears to no longer be available in 4.3, I think that this is a feature regression — there is currently no good way to add multiple nodes to groups in the editor (as far as I'm aware).

Besides updating the documentation to remove mentions of selecting multiple nodes (or how to do it, if it is in fact possible and Shift or Ctrl + clicking nodes in the Scene Tree is not the correct method), it may be best to either add the ability to select multiple nodes to add to groups, or to bring back the Manage Groups window from the last version.

**Comment by PixelSinger** [comment]
Remember that the group adds, removals, and calls all use a StringName for the group name. To make your groups more specific to its use (e.g. enemy group for scene A, enemy group for scene B, enemy group for children of node A, enemy group for children of node B) all you need do is append the unique name of the scene or node to the group name (e.g. "Enemies" + "Scene"+your_current_scene_number, or your_current_node_name + "Enemies") when using any group functions/methods. That way your groups can be specific to a scene/node while sharing a common name like Enemies or Spawners or Detectors or whatever suits the group being gathered together.

**Comment by kafkaphoenix** [comment]
How can we differentiate in code when creating a global or a scene group?
void add_to_group(group: StringName, persistent: bool = false)
doesn't have an option

**Reply by apanascx** [reply]
Groups are unique based on their name. The group "Enemies" is the same regardless if it's global or not. You can't have two groups called "Enemies". There is a Warning note in the documentation regarding this.

**Comment by LBdN** [comment]
to add to CrystalWarrior's point above, some "groups" are dynamic ( like the things within the camera frustum ) and it is often useful to get the intersection of a dynamic group with a specific static group. 
For example, in my game, any (wall) scene between the camera and the player character is turned invisible and its mesh children as assigned a dissolve shader. 
But it would useful to filter out among the (mesh) children that is dedicated to the minimap group or to lighting group. 
We can do it manually, but it is a shame we don't have a good way to do it.

**Comment by joethephish** [comment]
For features like this I’m always interested in the time complexity of the implementation behind the scenes. As far as I can tell, it’s:

* A dictionary type lookup when looking up a particular group name (eg when adding / removing members)
* Each group’s members are stored in a linear dynamic array, so adding, removing and checking for membership within a group is linear time complexity with the number of objects in the group

**Comment by leonardoraele** [comment]
How do you get the list of all existing groups? For example, to make an export field into a "enum_suggestion" where the user picks from a list of groups to avoid typo mismatches.

**Reply by thygrrr** [reply]
Project Settings -> Globals -> Groups, you can also reach that from code.

**Reply by leonardoraele** [reply]
How do you reach that from code?

**Reply by Bkjvk** [reply]
As far as I can see, you're supposed to know your groups. Do if you'd like to create some enum_suggestion I would recommend you create said enum and use - add_to_group method in _Ready method of said object with enum as a name.
Other way is define your groups in godot app and create enum that contain those names. 
(As enum I also consider list of string constans)

**Reply by leonardoraele** [reply]
@Bkjvk These options are only possible if you know the groups ahead of time. If, for example, you are writing a plugin, that's impossible.

**Reply by LawfulNeural** [reply]
Not sure how applicable this is to your use case, but adding my approach, in case anyone else finds it useful.

I create a global autoload named GroupRegistry. Within that global, every group is defined as `const GROUP_NAME: String = "group_name"`. You could also use enums or StringNames, I guess. Then I add stuff to groups within _ready() or _enter_tree() with `add_to_group(GroupRegistry.GROUP_NAME)`.

This has 2 perks:

1. you have a literal registry of every group in the project (whether in use or not)
2. you don't risk mistyping something, since you're referring to constants, not to hand-typed Strings

Not sure if there are any major downsides to this solution. Personally, I haven't encountered any. But if you can think of one, I'd love to hear it.

**Reply by leonardoraele** [reply]
@LawfulNeural This approach is great for your own projects, but if you are writing a plugin or an editir script that you intend to reuse in other projects, you can't rely on project-specific code like this.

**Comment by Wilker-uwu** [comment]
how i programmatically create a group? are groups created when named? how would i set it up for an addon?

**Comment by SnakeMcCloud** [comment]
Worth noting that it appears that `GetTree().CallGroup()` fails silently when the method called is an async task. I think an issue for this has been opened but figured it'd be worth posting here.

**Reply by Wilker-uwu** [reply]
same behavior as Signals then?
